File: solution/day23/src/test/java/org/craftedsw/tripservicekata/trip/GetTripsByUserShould.java
package org.craftedsw.tripservicekata.trip;

import org.craftedsw.tripservicekata.exception.UserNotLoggedInException;
import org.craftedsw.tripservicekata.user.User;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThat;
import static org.craftedsw.tripservicekata.user.UserBuilder.aUser;
import static org.junit.jupiter.api.DisplayNameGenerator.ReplaceUnderscores;
import static org.mockito.Mockito.when;

@DisplayNameGeneration(ReplaceUnderscores.class)
@ExtendWith(MockitoExtension.class)
class GetTripsByUserShould {
    private final User registeredUser = aUser().build();
    private TripService tripService;

    @Mock
    private TripRepository repositoryMock;

    @BeforeEach
    void setup() {
        tripService = new TripService(repositoryMock);
    }

    @Nested
    class return_An_Error {
        private final User guest = null;

        @Test
        void when_user_is_not_loggedIn() {
            assertThat(tripService.retrieveFriendTrips(registeredUser, guest).getCause())
                    .isInstanceOf(UserNotLoggedInException.class);
        }
    }

    @Nested
    class return_ {
        private final Trip lisbon = new Trip();
        private final Trip springfield = new Trip();
        private final User anotherUser = aUser().build();

        @Test
        void no_trips_when_logged_user_is_not_a_friend_of_the_target_user() {
            var aUserWithTrips = aUser()
                    .friendsWith(anotherUser)
                    .travelledTo(lisbon)
                    .build();

            assertThat(tripService.retrieveFriendTrips(aUserWithTrips, registeredUser).get())
                    .isEmpty();
        }

        @Test
        void all_the_target_user_trips_when_logged_user_and_target_user_are_friends() {
            var aUserWithTrips = aUser()
                    .friendsWith(anotherUser, registeredUser)
                    .travelledTo(lisbon, springfield)
                    .build();

            when(repositoryMock.findTripsByUser(aUserWithTrips))
                    .thenReturn(aUserWithTrips.getTrips());

            assertThat(tripService.retrieveFriendTrips(aUserWithTrips, registeredUser).get())
                    .hasSize(2)
                    .contains(lisbon, springfield);
        }
    }
}

File: solution/c#/Day23/Day23.Tests/Trips/GetTripsByUserShould.cs
using Day23.Exception;
using Day23.Trips;
using Day23.Users;
using FluentAssertions;
using FluentAssertions.LanguageExt;
using NSubstitute;
using Xunit;
using static Day23.Tests.Users.UserBuilder;

namespace Day23.Tests.Trips
{
    public class GetTripsByUserShould
    {
        private static readonly User RegisteredUser = AUser().Build();

        private readonly ITripRepository _repository = Substitute.For<ITripRepository>();
        private readonly TripService _tripService;

        private GetTripsByUserShould() => _tripService = new(_repository);

        public class Return : GetTripsByUserShould
        {
            private static readonly Trip Lisbon = new();
            private static readonly Trip Springfield = new();
            private static readonly User AnotherUser = AUser().Build();

            [Fact]
            public void No_Trips_When_Logged_User_Is_Not_A_Friend_Of_The_Target_User()
            {
                var aUserWithTrips = AUser()
                    .FriendsWith(AnotherUser)
                    .TravelledTo(Lisbon)
                    .Build();

                _tripService.RetrieveFriendTrips(
                        aUserWithTrips,
                        RegisteredUser)
                    .Should()
                    .BeSuccess(trips => trips.Should().BeEmpty());
            }

            [Fact]
            public void All_The_Target_User_Trips_When_Logged_User_Is_A_Friend_Of_The_Target_User()
                // Alternative syntax using Functional Extensions for our tests
                => AUser()
                    .FriendsWith(AnotherUser, RegisteredUser)
                    .TravelledTo(Lisbon, Springfield).Build()
                    .Let(aUserWithTrips =>
                    {
                        _repository.FindTripsByUser(aUserWithTrips)
                            .Returns(aUserWithTrips.Trips());

                        _tripService.RetrieveFriendTrips(
                                aUserWithTrips,
                                RegisteredUser)
                            .Should()
                            .BeSuccess(trips => trips.Should().ContainInOrder(Lisbon, Springfield));
                    });
        }

        public class ReturnAnError : GetTripsByUserShould
        {
            private readonly User? _guest = null;

            [Fact]
            public void When_User_Is_Not_LoggedIn()
                => _tripService
                    .RetrieveFriendTrips(RegisteredUser, _guest).Should()
                    .BeFail()
                    .Which.Should()
                    .BeOfType<UserNotLoggedInException>();
        }
    }
}